README.md
## stash-bot

Stash is a UCI-compliant chess engine developed from scratch. Preferably used
with a GUI like Arena, CuteChess, Fritz, etc. Currently rated around 2160 Elo.

## Files

The repository consists of the following files:

  * Readme.md, the file you are currently reading.
  * LICENSE, a text file containing the GNU General Public License version 3.
  * src, the directory containing all the source code + a Makefile that can be
    used to compile Stash on Unix-like systems (or Windows if you installed
    MinGW).
  * unix\_build.sh and windows\_build.sh, two shell scripts that can generate
    profiled builds for your architecture.
  * release\_build.sh, a shell script mainly intended for the contributors to
    automatically generate all binaries when creating releases.
  * Changelog, a text file listing all changes applied to this repository since
    version 9.

## UCI Parameters

Stash supports for now all these UCI options:

  * #### Hash
    Sets the hash table size in MB (defaults to 16).

  * #### Clear Hash
    Clears the hash table.

  * #### MultiPV
    Output the best N lines (principal variations) when searching.
	This option is still experimental (not completely UCI compliant),
	and so may not satisfy all GUIs. Leave at 1 for best performance.

  * #### Move Overhead
    Assumes a time delay of x milliseconds due to network and GUI overheads.
	Increase it if the engine often loses games on time. The default value
	of 20 ms should be reasonable for all chess GUIs.

  * #### Minimum Thinking Time
    Thinks at least x milliseconds per move. Useful for short time controls
	when the Move Overhead may cause the usable time to be zero. Too high values
	may cause the engine to flag bullet games.

## Frequently Asked Questions

  * #### How do I compile this project for my computer ?
    If you are on a Unix-like system, simply go to the src directory from a
	terminal and type:
	```
	make
	```
	Special architecture optimizations may be applied by typing:
	```
	make ARCH=arch_name
	```
	with `arch_name` being one of the following: x86-64, x86-64-modern or
	x86-64-bmi2.

  * #### I do not have a compiler on my machine: how do I do ?
    Compiled binaries for Linux and Windows are available from the "releases"
	page of the project. You can download the binary corresponding to your
	operating system and architecture.

  * #### Which binary should I choose ?
    The latest release should always be the best. The different architecture
	builds are:
	  - 64: generic 64-bit build. Should work on all 64-bit processors
	    of your operating system.

	  - x86_64: x86_64 build. Same as previous one, but enables use of the
	    `prefetch` instruction, which places RAM addresses in cache for faster
		access. Should work on almost all 64-bit processors.

	  - x86_64-modern: same as previous one, but also enables use of the
	    `popcnt` (Population Count) instruction. Should work on all > 2006
		64-bit processors.

	  - x86_64-bmi2: same as previous one, but also enables use of the `pext`
	    (Parallel Bit Extract) instruction. Should work on all AMD
		processors with Excavator arch or newer, and all Intel processors with
		Haswell arch or newer.


Changelog
2020-05-06	Morgan Houppin	<morganhouppin@gmail.com>

	* board.h: Added material key to board stack for endgame evaluation.
	* endgame.h: Added endgame specialized evaluations.
	* engine.h: Removed NMP_TrustScore (replaced by Victory Score).
	* info.h: Added TB hit counter (for now simply counting endgame positions
		reached in search).
	* score.h: Added Victory Score constant (10000 centipawns).
	* board/do_move.c: Added material key updating.
	* board/set_boardstack.c: Added material key initializing.
	* endgame/eval_draw.c: Added insufficient material draws + KNNK draw.
	* endgame/eval_kbbk.c: Added KBBK eval specialization.
	* endgame/eval_knbk.c: Added KNBK eval specialization.
	* endgame/eval_kqk.c: Added KQK eval specialization.
	* endgame/eval_kqkm.c: Added KQKM eval specialization.
	* endgame/eval_kqkr.c: Added KQKR eval specialization.
	* endgame/eval_krbk.c: Added KRBK eval specialization.
	* endgame/eval_krk.c: Added KRK eval specialization.
	* endgame/init_endgame_table.c: Added endgame table initialization.
	* engine/engine_go.c: Added TB hit display.
	* engine/evaluate.c: Retired insufficient material checking, added endgame
		specialized eval probing.
	* engine/out_of_time.c: Just retired a newline.
	* engine/search_bestmove.c: Restored LMR computation, improved fail-low
		value returns.
	* main/main.c: Added endgame initialization call.
	* misc/globals.c: Added g_tbhits symbol.
	* pawns/evaluate_pawns.c: Added Doubled/Isolated Pawn Penalty.

2020-05-04	Morgan Houppin	<morganhouppin@gmail.com>

	* engine.h: Added static evaluation to search stack.
	* imath.h: Added min, max and clamp helpers.
	* movelist.h: Added pseudo-legal move generator.
	* tt.h: Added static evaluation to hash entries.
	* engine/search_bestmove.c: Replaced legal by pseudo-legal move generators,
		added improving eval check for Null Move Pruning.
	* movelist/generate_instable.c: Removed legality checking (now verified
		during search).
	* movelist/generate_pseudo.c: Added pseudo-legal move generator.
	* tt/tt_save.c: Added static evaluation to hash saving.

2020-04-28	Morgan Houppin	<morganhouppin@gmail.com>

	* uci/uci_d.c: Now printing static eval and Zobrist key.
	* uci/uci_position.c: Fixed global access race on board.
	Written Changelog from version 9 to actual version.
	Changed all file headers.

2020-04-26	Morgan Houppin	<morganhouppin@gmail.com>

	* engine/evaluate.c: Added pawn structure and piece mobility.
	* uci/uci_uci.c: Changed version to 15.

2020-04-25	Morgan Houppin	<morganhouppin@gmail.com>

	* engine/search_bestmove.c: Increased LMR with depth.

2020-04-24	Morgan Houppin	<morganhouppin@gmail.com>

	* engine/evaluate.c: Added insufficient material rescoring.
	* engine/search_bestmove.c: Added killer heuristic.
	* movelist/generate_move_values.c: Added killer moves to quiet scoring.

2020-04-18	Morgan Houppin	<morganhouppin@gmail.com>

	* engine/search_bestmove.c: Increased NMP reduction with depth.

2020-04-16	Morgan Houppin	<morganhouppin@gmail.com>

	* main/main.c: Init position as startpos to avoid crashes.

2020-04-14	Morgan Houppin	<morganhouppin@gmail.com>

	* engine/search_bestmove.c: Added qsearch TT probing/writing.

2020-04-09	Morgan Houppin	<morganhouppin@gmail.com>

	* engine/engine_go.c: Added 'go mate x' command support.
	* uci/uci_uci.c: Changed version to 14.

2020-04-08	Morgan Houppin	<morganhouppin@gmail.com>

	* engine/search_bestmove.c: Implemented razoring.

2020-04-07	Morgan Houppin	<morganhouppin@gmail.com>

	* engine/search_bestmove.c: Qsearch eval pruning only when not in check.

2020-04-05	Morgan Houppin	<morganhouppin@gmail.com>

	* engine/search_bestmove.c: Splitted search between PV/non-PV nodes.

2020-04-01	Morgan Houppin	<morganhouppin@gmail.com>

	* engine/engine_go.c: Added 'go nodes/depth x' command correct support.
	* uci/uci_uci.c: Changed version to 13.

2020-03-31	Morgan Houppin	<morganhouppin@gmail.com>

	* engine/engine_go.c: Fixed LMR depth activation.

2020-03-26	Morgan Houppin	<morganhouppin@gmail.com>

	* movelist/generate_move_values.c: Added SEE to capture scoring.

2020-03-24	Morgan Houppin	<morganhouppin@gmail.com>

	* engine/engine_go.c: Added 'go perft x' command support.
	* engine/search_bestmove.c: Fixed TT writing for partial bounds.
	* uci/uci_bench.c: Fixed issues with format strings.

2020-03-23	Morgan Houppin	<morganhouppin@gmail.com>

	* engine/search_bestmove.c: Added NMP mate validation.

2020-03-19	Morgan Houppin	<morganhouppin@gmail.com>

	* engine/search_bestmove.c: Added Late Move Reductions.
	* uci/uci_uci.c: Changed version to 12.

2020-03-16	Morgan Houppin	<morganhouppin@gmail.com>

	* engine/search_bestmove.c: Tuned NMP for zugzwang checking.

2020-03-13	Morgan Houppin	<morganhouppin@gmail.com>

	* engine/search_bestmove.c: Added Null Move Pruning.

2020-03-11	Morgan Houppin	<morganhouppin@gmail.com>

	* engine/search_bestmove.c: Disabled TT probing of past search vlaues.
	* uci/uci_uci.c: Changed version to 11.

2020-03-07	Morgan Houppin	<morganhouppin@gmail.com>

	Added Transposition Tables.

2020-03-03	Morgan Houppin	<morganhouppin@gmail.com>

	* engine/search_bestmove.c: Added Quiescence search.
	* uci/uci_uci.c: Changed version to 10.

2020-02-27	Morgan Houppin	<morganhouppin@gmail.com>

	Replaced rand() by a xor-shift function.
	* bitboard.h: Added conditional including for immintrin header.
	Reverted PST scores (midgame<->endgame)
	* uci/uci_uci.c: Changed version to 9.


include/board.h
/* ************************************************************************** */
/*                                                          LE - /            */
/*                                                              /             */
/*   board.h                                          .::    .:/ .      .::   */
/*                                                 +:+:+   +:    +:  +:+:+    */
/*   By: stash <stash@student.le-101.fr>            +:+   +:    +:    +:+     */
/*                                                 #+#   #+    #+    #+#      */
/*   Created: 2020/02/18 15:13:19 by stash        #+#   ##    ##    #+#       */
/*   Updated: 2020/02/23 20:58:47 by stash       ###    #+. /#+    ###.fr     */
/*                                                         /                  */
/*                                                        /                   */
/* ************************************************************************** */

#ifndef BOARD_H
# define BOARD_H

# include <stdbool.h>
# include "bitboard.h"
# include "castling.h"
# include "color.h"
# include "hashkey.h"
# include "inlining.h"
# include "move.h"
# include "piece.h"
# include "square.h"
# include "psq_score.h"

typedef struct boardstack_s	boardstack_t;

struct	boardstack_s
{
	int				castlings;
	int				rule50;
	int				plies_from_null_move;
	square_t		en_passant_square;
	hashkey_t		board_key;
	bitboard_t		checkers;
	piece_t			captured_piece;
	boardstack_t	*prev;
	bitboard_t		king_blockers[COLOR_NB];
	bitboard_t		pinners[COLOR_NB];
	bitboard_t		check_squares[PIECETYPE_NB];
	int				repetition;
};

typedef struct	board_s
{
	piece_t			table[SQUARE_NB];
	bitboard_t		piecetype_bits[PIECETYPE_NB];
	bitboard_t		color_bits[COLOR_NB];
	int				piece_count[PIECE_NB];
	square_t		piece_list[PIECE_NB][16];
	int				piece_index[SQUARE_NB];
	int				castling_mask[SQUARE_NB];
	square_t		castling_rook_square[CASTLING_NB];
	bitboard_t		castling_path[CASTLING_NB];
	int				ply;
	color_t			side_to_move;
	scorepair_t		psq_scorepair;
	boardstack_t	*stack;
	bool			chess960;
}				board_t;

bool			board_legal(const board_t *board, move_t move);
bool			move_gives_check(const board_t *board, move_t move);

bitboard_t		attackers_list(const board_t *board, square_t s,
				bitboard_t occupied);

bitboard_t		slider_blockers(const board_t *board, bitboard_t sliders,
				square_t square, bitboard_t *pinners);

void			set_castling(board_t *board, color_t color,
				square_t rook_square);

void			set_boardstack(board_t *board, boardstack_t *stack);
void			set_check(board_t *board, boardstack_t *stack);

void			board_set(board_t *board, char *fen, bool is_chess960,
				boardstack_t *bstack);

void			undo_move(board_t *board, move_t move);
void			do_move_gc(board_t *board, move_t move, boardstack_t *stack,
				bool gives_check);

void			do_castling(board_t *board, color_t us, square_t king_from,
				square_t *king_to, square_t *rook_from, square_t *rook_to);

void			undo_castling(board_t *board, color_t us, square_t king_from,
				square_t *king_to, square_t *rook_from, square_t *rook_to);

void			do_null_move(board_t *board);
void			undo_null_move(board_t *board);

bool			is_draw(const board_t *board, int ply);

INLINED	bool	empty_square(const board_t *board, square_t square)
{
	return (board->table[square] == NO_PIECE);
}

INLINED	piece_t	piece_on(const board_t *board, square_t square)
{
	return (board->table[square]);
}

INLINED piece_t	moved_piece(const board_t *board, move_t move)
{
	return (board->table[move_from_square(move)]);
}

INLINED bitboard_t	board_pieces(const board_t *board, piecetype_t pt1,
					piecetype_t pt2)
{
	return (board->piecetype_bits[pt1] | board->piecetype_bits[pt2]);
}

INLINED bitboard_t	board_colored_pieces(const board_t *board, color_t color,
					piecetype_t pt1, piecetype_t pt2)
{
	return ((board->piecetype_bits[pt1] | board->piecetype_bits[pt2])
		& board->color_bits[color]);
}

INLINED bitboard_t	king_moves(square_t square)
{
	return (PseudoMoves[KING][square]);
}

INLINED bitboard_t	knight_moves(square_t square)
{
	return (PseudoMoves[KNIGHT][square]);
}

INLINED bitboard_t	pawn_moves(square_t square, color_t color)
{
	return (PawnMoves[color][square]);
}

INLINED bitboard_t	bishop_moves(const board_t *board, square_t square)
{
	return (bishop_move_bits(square, board->piecetype_bits[ALL_PIECES]));
}

INLINED bitboard_t	rook_moves(const board_t *board, square_t square)
{
	return (rook_move_bits(square, board->piecetype_bits[ALL_PIECES]));
}

INLINED bitboard_t	queen_moves(const board_t *board, square_t square)
{
	return (bishop_move_bits(square, board->piecetype_bits[ALL_PIECES])
		| rook_move_bits(square, board->piecetype_bits[ALL_PIECES]));
}

INLINED bitboard_t	piece_moves(piecetype_t piecetype, square_t square,
					bitboard_t occupied)
{
	switch (piecetype)
	{
		case KNIGHT:
			return (PseudoMoves[KNIGHT][square]);

		case BISHOP:
			return (bishop_move_bits(square, occupied));

		case ROOK:
			return (rook_move_bits(square, occupied));

		case QUEEN:
			return (bishop_move_bits(square, occupied)
				| rook_move_bits(square, occupied));

		default:
			__builtin_unreachable();
			return (0);
	}
}

INLINED bitboard_t	attackers_to(const board_t *board, square_t square)
{
	return (attackers_list(board, square, board->piecetype_bits[ALL_PIECES]));
}

INLINED bool		castling_blocked(const board_t *board, int castling)
{
	return (board->piecetype_bits[ALL_PIECES] & board->castling_path[castling]);
}

INLINED void		put_piece(board_t *board, piece_t piece, square_t square)
{
	const bitboard_t	bitsquare = square_bit(square);

	board->table[square] = piece;
	board->piecetype_bits[ALL_PIECES] |= bitsquare;
	board->piecetype_bits[type_of_piece(piece)] |= bitsquare;
	board->color_bits[color_of_piece(piece)] |= bitsquare;
	board->piece_index[square] = board->piece_count[piece]++;
	board->piece_list[piece][board->piece_index[square]] = square;
	board->piece_count[create_piece(color_of_piece(piece), ALL_PIECES)]++;
	board->psq_scorepair += PsqScore[piece][square];
}

INLINED void		move_piece(board_t *board, square_t from, square_t to)
{
	piece_t		piece = board->table[from];
	bitboard_t	move_bits = square_bit(from) | square_bit(to);

	board->piecetype_bits[ALL_PIECES] ^= move_bits;
	board->piecetype_bits[type_of_piece(piece)] ^= move_bits;
	board->color_bits[color_of_piece(piece)] ^= move_bits;
	board->table[from] = NO_PIECE;
	board->table[to] = piece;
	board->piece_index[to] = board->piece_index[from];
	board->piece_list[piece][board->piece_index[to]] = to;
	board->psq_scorepair += PsqScore[piece][to] - PsqScore[piece][from];
}

INLINED void		remove_piece(board_t *board, square_t square)
{
	piece_t				piece = board->table[square];
	const bitboard_t	bitsquare = square_bit(square);

	board->piecetype_bits[ALL_PIECES] ^= bitsquare;
	board->piecetype_bits[type_of_piece(piece)] ^= bitsquare;
	board->color_bits[color_of_piece(piece)] ^= bitsquare;

	square_t	last_square = board->piece_list[piece][
		--board->piece_count[piece]];

	board->piece_index[last_square] = board->piece_index[square];
	board->piece_list[piece][board->piece_index[last_square]] = last_square;
	board->piece_list[piece][board->piece_count[piece]] = SQ_NONE;
	board->piece_count[create_piece(color_of_piece(piece), ALL_PIECES)]--;
	board->psq_scorepair -= PsqScore[piece][square];
}

INLINED void		do_move(board_t *board, move_t move, boardstack_t *stack)
{
	do_move_gc(board, move, stack, move_gives_check(board, move));
}

#endif
